src/int/mod.rs

```diff
@@ -178,54 +279,110 @@ macro_rules! int_impl {
     };
 }
 
+int_impl!(isize, usize, usize::MAX.count_ones());
+int_impl!(i8, u8, 8);
 int_impl!(i16, u16, 16);
 int_impl!(i32, u32, 32);
 int_impl!(i64, u64, 64);
 int_impl!(i128, u128, 128);
 
-/// Trait to convert an integer to/from smaller parts
-pub(crate) trait LargeInt: Int {
-    type LowHalf: Int;
-    type HighHalf: Int;
+/// Trait for integers twice the bit width of another integer. This is implemented for all
+/// primitives except for `u8`, because there is not a smaller primitive.
+#[doc(hidden)]
+pub trait DInt: Int {
+    /// Integer that is half the bit width of the integer this trait is implemented for
+    type H: HInt<D = Self> + Int;
+
+    /// Returns the low half of `self`
+    fn lo(self) -> Self::H;
+    /// Returns the high half of `self`
+    fn hi(self) -> Self::H;
+    /// Returns the low and high halves of `self` as a tuple
+    fn lo_hi(self) -> (Self::H, Self::H);
+}
 
-    fn low(self) -> Self::LowHalf;
-    fn low_as_high(low: Self::LowHalf) -> Self::HighHalf;
-    fn high(self) -> Self::HighHalf;
-    fn high_as_low(low: Self::HighHalf) -> Self::LowHalf;
-    fn from_parts(low: Self::LowHalf, high: Self::HighHalf) -> Self;
+/// Trait for integers half the bit width of another integer. This is implemented for all
+/// primitives except for `u128`, because it there is not a larger primitive.
+#[doc(hidden)]
+pub trait HInt: Int {
+    /// Integer that is double the bit width of the integer this trait is implemented for
+    type D: DInt<H = Self> + Int;
+
+    /// Widens (using default extension) the integer to have double bit width
+    fn widen(self) -> Self::D;
+    /// Widens (zero extension only) the integer to have double bit width. This is needed to get
+    /// around problems with associated type bounds (such as `Int<Othersign: DInt>`) being unstable
+    fn zero_widen(self) -> Self::D;
+    /// Widens the integer to have double bit width and shifts the integer into the higher bits
+    fn widen_hi(self) -> Self::D;
+    /// Constructs a double bit width integer using lower and higher parts
+    fn from_lo_hi(lo: Self, hi: Self) -> Self::D;
```

Comment:
I feel the opposite. I think it should be consistent with the other methods in `DInt` which all return `Self::D`, and all the methods in `HInt` return `Self::H`. Consider if we added a `widening_mul` method to the standard library (and I named it `widen_mul` on this trait to avoid colliding with [this rfc](https://github.com/rust-lang/rfcs/pull/2417) if it were ever implemented). You would widen multiply a pair of `u64`s into a `u128`, so it would be called like `u64.widening_mul(u64) -> u128`. In a similar way, I want the signature of `from_lo_hi` to be used like `u128::from_lo_hi(u64, u64) -> u128` (and I use this particular signature to mirror the tuple from `HInt::lo_hi`).

Reply:
Currently `from_lo_hi` is called as `u64::from_lo_hi(u64, u64) -> u128` which doesn't make sense. You want `u128:from_lo_hi` instead of `u64::from_lo_hi` which requires the method to be on the `DInt` trait.

Reply:
You're right, I don't know why I didn't see this before. It eliminates some code size in practice also